### PR TITLE
VZ-5621: propagate CA certs between admin and managed clusters when they change

### PR DESCRIPTION
**Pull request by wmhopkins**
# Description

Add code to the MC agent in the application operator that checks whether the admin and/or local cluster CA certs have changed -- either because the expired and were reissued, or because DNS/certificate configuration has changed -- and, if so, synchronizes the changes. A new admin cert is pulled down and updated in the cluster-registration-secret; a new manged cluster CA is pushed to the VMC CA secret on the admin cluster.

Also added a new controller in the platform operator to create a new secret in verrazzano-mc namespace, copy the verrazzano-tls ca.crt value (admin ca bundle) to the new secret, and keep it updated if the admin CA changes. This is so that managed clusters can read the secret -- they have no access to verrazzano-system namespace.

Fixes VZ-5621

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/mcagent/mcagent_clusterca.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2021, 2022, Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline application-operator/mcagent/mcagent_clusterca.go:1]: the "2021," can be removed for this new file

File: application-operator/mcagent/mcagent_clusterca.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		s.Log.Errorf("Error syncing Local Cluster CA: %v", err)
+	}
+	return nil
+}
+
+// syncAdminClusterCA - synchronize the admin cluster CA cert -- update local copy if admin CA changes
+func (s *Syncer) syncAdminClusterCA() error {
+
+	s.Log.Info("Syncing AdminClusterCA ...")
```
Comment (mgianatagh) [inline application-operator/mcagent/mcagent_clusterca.go:35]: Maybe these Info messages should be "progress" messages to have less noise in the log

File: application-operator/mcagent/mcagent_clusterca.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	}, &registrationSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	s.Log.Info("Got local cluster registration secret")
+
+	// Update the local cluster registration secret if the admin CA certs are different
+	if !secretsEqualTrimmedWhitespace(registrationSecret.Data["ca-bundle"], adminCASecret.Data["ca.crt"]) {
+		s.Log.Info("CAs are different -- updating")
+		s.Log.Info("Local registration Secret: %v", registrationSecret.Data["ca-bundle"])
```
Comment (mgianatagh) [inline application-operator/mcagent/mcagent_clusterca.go:62]: Is the logging of the registration secret a temporary measure while WIP?

File: platform-operator/controllers/clusterca/clusterca_controller.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2021, 2022, Oracle and/or its affiliates.
```
Comment (mgianatagh) [inline platform-operator/controllers/clusterca/clusterca_controller.go:1]: 2021 can be removed from the copyright

Comment (wmhopkins) [reply platform-operator/controllers/clusterca/clusterca_controller.go:1]: done

File: application-operator/mcagent/mcagent_clusterca.go (hunk 3 above)
Comment (wmhopkins) [reply application-operator/mcagent/mcagent_clusterca.go:62]: Yes, will remove before merging.

File: application-operator/mcagent/mcagent_clusterca.go (hunk 1 above)
Comment (wmhopkins) [reply application-operator/mcagent/mcagent_clusterca.go:1]: done

File: platform-operator/controllers/clusterca/clusterca_controller.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// VerrazzanoAdminCAReconciler reconciles a the Admin CA bundle by copying it
+// to a secret in the verrazzano-mc namespace, so managed clusters can pull it.
+type VerrazzanoAdminCAReconciler struct {
```
Comment (michael-cico) [inline platform-operator/controllers/clusterca/clusterca_controller.go:24]: I wonder if we can just make this a general secrets reconciler.  i.e., we could use this to sync things like updated OCI secrets, or image pull secrets across the system namespaces.  Although there are other tools we could use for that.

Not saying that we should do all that now, but we could generalize the naming etc of this reconciler and use it to do other things like the things mentioned above later.

Comment (wmhopkins) [reply platform-operator/controllers/clusterca/clusterca_controller.go]: done.

File: platform-operator/controllers/secrets/secrets_controller.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+func (r *VerrazzanoSecretsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	// The only secret we care about (for now) is the verrazzano ingress tls secret (verrazzano-tls)
+	if req.Name != constants.VerrazzanoIngressSecret || req.Namespace != constants.VerrazzanoSystemNamespace {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the verrazzano ingress secret
+	caSecret := corev1.Secret{}
```
Comment (michael-cico) [inline platform-operator/controllers/secrets/secrets_controller.go:48]: Have we tried this with a customer-provided CA secret?  i.e., is `ca.crt` populated in that case?

**Review by michael-cico — APPROVED** [review]

**Pull request merged**